wave_lang/kernel/wave/scheduling/multi_buffering.py

```diff
@@ -23,10 +21,78 @@
     Write,
     get_custom,
 )
+import copy
 from ..utils.mapping_utils import get_dict_with_updated_key
+from typing import Optional
 
 
-def multi_buffer(trace: CapturedTrace):
+def heuristically_multi_buffer_shared_memory(
+    new_node: CustomOp,
+    induction_variable: IndexSymbol,
+    reduction: Iterate,
+    multi_buffer_count: Optional[int] = None,
+):
+    """
+    This function adjusts the offset to a shared memory buffer to multibuffer.
+    This enables the removal of LDS barriers and enables greater parallelism
+    during software pipelining and schedule reordering techniques
+
+    The induction variable used in this function is the induction variable before
+    it's remapped during schedule reordering. (See update_node_index in loop_reconstuction.py)
+
+    This enables index to be set to buffer_size * (original_induction_variable % multi_buffer_count)
+    Regardless of current node stage, because it will be remapped correctly later depending on if
+    prologue, epilogue, or kernel stage
+    """
+    assert multi_buffer_count is not None
+    is_read_write = isinstance(new_node, Read | Write)
+    original_buffer = (
+        get_custom(new_node.memory) if is_read_write else get_custom(new_node.dst)
+    )
+    reduction_axis = reduction.axis
+    reduction_dim_indices = [
+        i for i, dim in enumerate(original_buffer.shape) if dim == reduction_axis
+    ]
+    for i, dim in enumerate(original_buffer.shape):
+        new_node_idx = copy.deepcopy(
+            new_node.index if is_read_write else new_node.dst_index
+        )
+        if i in reduction_dim_indices or dim not in new_node_idx:
+            continue
+        block_size = original_buffer.distributed_shape[i]
+        which_buffer = induction_variable % multi_buffer_count
+        offset = block_size * which_buffer
+        new_node_idx[dim].start += offset
+
+        # Update the mapping for the operation as the keys for the
+        # mapping have to match the shape of memory location the
+
+        # operation reads from / writes to, which we change below.
+        new_node_idx = get_dict_with_updated_key(
+            new_node_idx, dim, dim * multi_buffer_count
+        )
+        if is_read_write:
+            new_node.index = new_node_idx
+        else:
+            new_node.dst_index = new_node_idx
+            return
```

Comment:
Tiny tiny suggestion for clarity and brevity

```suggestion
        if not is_read_write:
            new_node.dst_index = new_node_idx
            return
            
        new_node.index = new_node_idx
```

Reply:
LGTM